ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java

```diff
@@ -118,47 +121,55 @@ public synchronized void trackSerdeOp(
    */
-  public SchemaInfo getTopicInfo(final String topicName) {
+  public MultiSchemaInfo getTopicInfo(final String topicName) {
+    final Set<SchemaInfo> keySchemasInfo = getTopicSchemas(topicName, IS_KEY_SCHEMA);
+    final Set<SchemaInfo> valueSchemasInfo = getTopicSchemas(topicName, IS_VALUE_SCHEMA);
+
+    return new MultiSchemaInfo(keySchemasInfo, valueSchemasInfo);
+  }
+
+  /**
+   * Returns a list of schemas and serde formats detected on the specified topicName.
+   * The isKeySchema is a boolean value that specifies whether to look at the key schema (True)
+   * or at the value schema (False).
+   * </p>
+   * Each schema information is tracked by the loggers the topicName has. The logger is
+   * a changelog topic. Some topics may have multiple schemas and serde formats if the topicName
+   * is used with stream-stream joins and foreign key joins.
+   */
+  private Set<SchemaInfo> getTopicSchemas(final String topicName,
+                                          final boolean isKeySchema) {
+    // Look at the different changelog topics that tne topicName uses. There might be multiple
+    // changelog topics if the topicName was used by joins, which contains internal state stores
+    // with a changelog per store.
     final Map<Boolean, Set<String>> kvLoggerNames = topicsToLoggers.get(topicName);
     if (kvLoggerNames == null) {
       throw new IllegalArgumentException("Unknown topic: " + topicName);
     }
 
-    final Set<String> keyLoggerNames = kvLoggerNames.getOrDefault(true, ImmutableSet.of());
-    if (keyLoggerNames.size() != 1) {
-      throw new IllegalStateException("Multiple key logger names registered for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "loggers: " + keyLoggerNames
-      );
-    }
-
-    final Set<String> valueTopicNames = kvLoggerNames.getOrDefault(false, ImmutableSet.of());
-    if (valueTopicNames.size() != 1) {
-      throw new IllegalStateException("Multiple value logger names registered for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "loggers: " + valueTopicNames
-      );
-    }
+    final Map<SchemaInfo, Set<String>> schemaToLoggers = new HashMap<>();
 
-    final SchemaInfo keyInfo = loggerToSchemas.get(Iterables.getOnlyElement(keyLoggerNames));
-    final SchemaInfo valueInfo = loggerToSchemas.get(Iterables.getOnlyElement(valueTopicNames));
-    if (keyInfo == null || valueInfo == null) {
-      throw new IllegalStateException("Incomplete schema info for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "keyInfo: " + keyInfo
-          + "valueInfo: " + valueInfo
-      );
+    // Look at the changelogs linked to the key or value serde, and get all schema and formats
+    // detected.
+    for (final String loggerName : kvLoggerNames.getOrDefault(isKeySchema, ImmutableSet.of())) {
+      final SchemaInfo keyInfo = loggerToSchemas.get(loggerName);
```

Comment:
Why is this called `keyInfo`? It might also represent info about the value instead, right?

Reply:
I'll change it.